diesel/src/mysql/types/json.rs

```diff
@@ -0,0 +1,20 @@
+use crate::mysql::{Mysql, MysqlValue};
+use crate::deserialize::{self, FromSql};
+use crate::serialize::{self, IsNull, Output, ToSql};
+use crate::sql_types;
+use std::io::prelude::*;
+
+impl FromSql<sql_types::Json, Mysql> for serde_json::Value {
+    fn from_sql(value: Option<MysqlValue<'_>>) -> deserialize::Result<Self> {
+        let value = not_none!(value);
+        serde_json::from_slice(value.as_bytes()).map_err(|_| "Invalid Json".into())
+    }
+}
+
+impl ToSql<sql_types::Json, MysqlValue> for serde_json::Value {
```

Comment:
I have been meaning to send a PR. I want to do `impl ToSql<Json> for String` also for Pg. Can we do that here too?

Reply:
Our rules for adding `ToSql` imps in diesel itself are roughly :
1. Its a commonly used type mapping
2. It cannot fail because of invalid values. All (or almost all) valid values of the rust type must be representable by the corresponding sql type.

I cannot answer the second point with: "Yes that assumption is true" as an arbitrary string is not a valid json value for the general case. Therefore such impl does not exist yet and I would prefer to keep it that way.

Reply:
I wanted to add more support for the Json types which is why I was looking at this. I am just concerned with the performance because the deserialization and serialization happening all the time. Since I was dealing with non-user provided JSON, I wanted to map it to string with an `unwrap`. I understand your point though, but I would still say that we should add String conversions (thought heavily discouraged in Docs) with `unwrap`.

Reply:
Can you provide an example of what you are exactly trying to achieve in the end. In theory diesel already exposes some helper impls (although a bit more low level) than this one.

Reply:
I want to use the postgres `@>` operator to compare json without me needing to understand them. Let's say I have a table `rules` with json column `rule`. I want to query the following:

```sql
select * from rules where '{"a": "read", "b": 2}'::jsonb @> rule;
```

Where the `{"a": "read", "b": 2}` is user provided but validated externally and confirmed as JSON by a faster json checker (and not deserialized because I don't need the data).

So, 2 things here:

1. I would never want to store `rule` as `serde::Value` in my code.
2. I would never want to pass a certain json value as `serde::Value` since I do not want to serialize/deserialize it.

I hope I was able to explain what I mean here.

Reply:
As a general note: Diesel cannot reason about which things are externally validated. In such cases we generally choose to enforce an internal validation as well as we cannot assume that such an external validation has really happened. For this case this means I'm quite unwilling to add a `ToSql<Json, DB> for String` impl here as that will clearly blow up for some use cases. 
That written for a third party crate there are several possibilities to do something like that without touching `serde::Value`:

* Implement a custom query node for inserting the provided json as into the query sql
* Use a new type wrapper around `String` and implement `ToSql<Json, DB>` for this 
* Implement a cast query node, send the value as plain `String`/`Text` and cast to `jsonb` afterwards (technically that's what's already done in your example query)
* Introduce a new sql type that is basically a crate local version `Json` and implement

Beside of that diesel does currently not implement support for json operators. I think it's quite straight forward to define them via `diesel::infix_operator!` in a third party crate, but adding support for a common set of such operators is definitively something I would like to see as part of diesel.

Reply:
Thanks, I didn't know about the insert query node and cast query node. Still  working on understanding diesel more. Your answer addresses my problem.

I will definitely add support for the json operators in diesel later. Just trying to get the full overview first.

Reply:
If there is anything unclear about some internals, just keep asking questions. Unfortunately not everything is documented quite well…

About the operators: See this [implementation](https://github.com/diesel-rs/diesel/blob/4bd9d1948b3d1647dfb9168ab837ab08beb54d68/diesel/src/pg/expression/expression_methods.rs#L136-L300) of the postgres array operators implementation as example for built in support in diesel. Something like that also works in a third party crate.